### PR TITLE
EVG-6620 Support EBS volumes for spawn hosts

### PR DESCRIPTION
**Pull request by tzembo**
Still missing a few things (and needs more testing) but wanted to put this up for people to take a look. Created a new collection and struct in the host package for volumes, along with new commands/routes that allow users to create and delete volumes and attach them to hosts.

> evergreen host create-volume --type TYPE --size SIZE
> evergreen host delete-volume --id VOLUME_ID
> evergreen host modify --host HOST_ID  --attach VOLUME_ID
> evergreen host modify --host HOST_ID  --detach VOLUME_ID

Let you know if you have thoughts about these ^

Currently have hard-coded support for only EBS volumes in the availability zone "us-east-1a" -- although eventually we want to allow people to create volumes wherever their hosts their hosts are running.

### CODE REVIEW

File: model/host/volume.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"github.com/evergreen-ci/evergreen/db"
+	adb "github.com/mongodb/anser/db"
+	"go.mongodb.org/mongo-driver/bson"
+)
+
+type Volume struct {
+	ID        string `bson:"_id" json:"id"`
+	CreatedBy string `bson:"created_by" json:"created_by"`
```
Comment (tychoish) [inline model/host/volume.go]: is this a user name or a host id?

File: rest/route/host_spawn.go [hunk 2]
```diff
@@ -344,6 +352,142 @@ func (h *hostStartHandler) Run(ctx context.Context) gimlet.Responder {
...
+	}
+
+	// Only allow users to delete their own volumes
+	if u.Id != volume.CreatedBy {
+		return gimlet.MakeJSONErrorResponder(gimlet.ErrorResponse{
+			StatusCode: http.StatusUnauthorized,
+			Message:    errors.Errorf("not authorized to delete volume '%s'", volume.ID).Error(),
+		})
+	}
+	// TODO: Allow different providers/regions
```
Comment (tychoish) [inline rest/route/host_spawn.go]: +1

File: rest/route/service.go [hunk 3]
```diff
@@ -92,6 +92,8 @@ func AttachHandler(app *gimlet.APIApp, opts HandlerOpts) {
 	app.AddRoute("/hosts/{host_id}/logs/error").Version(2).Get().RouteHandler(makeContainerLogsRouteManager(sc, true))
 	app.AddRoute("/hosts/{task_id}/create").Version(2).Post().Wrap(checkTask).RouteHandler(makeHostCreateRouteManager(sc))
 	app.AddRoute("/hosts/{task_id}/list").Version(2).Get().Wrap(checkTask).RouteHandler(makeHostListRouteManager(sc))
+	app.AddRoute("/volumes").Version(2).Post().Wrap(checkUser).RouteHandler(makeCreateVolume(sc))
+	app.AddRoute("/volumes/{volume_id}").Version(2).Wrap(checkUser).Delete().RouteHandler(makeDeleteVolume(sc))
```
Comment (tychoish) [inline rest/route/service.go]: should we also add a list volumes? 

should these be hosts/volumes methods or no?

**Comment by ablack12** [comment]
I agree we should have a 
`evergreen host list-volumes` (that gives the size, type, ID, and host if applicable).
However, if that's out of scope of this ticket given your end date I'm fine making that a separate ticket.

File: model/host/volume.go (hunk 1 above)
Comment (tzembo) [reply model/host/volume.go]: CreatedBy is a user name.

File: model/host/volume.go [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		},
+	)
+}
+
+// Find a volume by its ID field
+func FindVolumeByID(id string) (*Volume, error) {
+	volume := &Volume{}
+	err := db.FindOneQ(
+		VolumesCollection,
+		db.Query(bson.D{{Key: VolumeIDKey, Value: id}}),
```
Comment (ablack12) [inline model/host/volume.go]: I'm not sure why we have it worded like this in some places, you should just be able to do the simpler `db.Query(bson.M{VersionIdKey: id})`

File: rest/route/service.go (hunk 3 above)
Comment (tzembo) [reply rest/route/service.go]: I don't see a huge reason why they can't be hosts/volumes. Was worried that we would have a POST route on hosts/volumes that would conflict with a POST route like hosts/host_id (which we don't have).

File: operations/host_spawn.go [hunk 5]
```diff
@@ -344,6 +362,103 @@ func hostStart() cli.Command {
...
+		sizeFlag = "size"
+		typeFlag = "type"
+	)
+
+	return cli.Command{
+		Name:  "create-volume",
+		Usage: "create a volume for spawn hosts",
+		Flags: []cli.Flag{
+			cli.StringFlag{
+				Name:  joinFlagNames(zoneFlag, "z"),
```
Comment (ablack12) [inline operations/host_spawn.go]: If we aren't supporting availability zones in this ticket then I think we should simply remove the flag until we do.

File: cloud/ec2.go [hunk 6]
```diff
@@ -613,6 +616,30 @@ func (m *ec2Manager) extendExpiration(ctx context.Context, h *host.Host, extensi
 	return errors.Wrapf(h.SetExpirationTime(h.ExpirationTime.Add(extension)), "error extending expiration time in db for '%s'", h.Id)
 }
 
+func (m *ec2Manager) attachVolume(ctx context.Context, h *host.Host, volume string) error {
```
Comment (ablack12) [inline cloud/ec2.go]: Are we checking somewhere that the volume isn't attached to another host? Better not to rely on AWS for this.

File: rest/route/service.go (hunk 3 above)
Comment (tzembo) [reply rest/route/service.go]: Created https://jira.mongodb.org/browse/EVG-6841

File: model/host/host.go [hunk 7]
```diff
@@ -1820,3 +1822,44 @@ func (h *Host) MarkShouldExpire(expireOnValue string) error {
...
+		return nil
+	}
+
+	return UpdateOne(
+		bson.M{
+			IdKey: h.Id,
+		},
+		bson.M{
+			"$set": bson.M{
+				VolumeIDsKey: h.VolumeIDs,
```
[the remark this answers is not in the document]
Comment (tzembo) [reply model/host/host.go]: Used the $pull operator, but need to iterate through the list anyway to remove the volume from the host document's volume ID slice.

File: cloud/ec2.go [hunk 8]
```diff
@@ -631,21 +658,13 @@ func (m *ec2Manager) ModifyHost(ctx context.Context, h *host.Host, opts host.Hos
...
-	if len(opts.AddInstanceTags) > 0 {
-		catcher.Add(m.addTags(ctx, h, opts.AddInstanceTags))
-	}
-	if opts.NoExpiration != nil {
-		catcher.Add(m.setNoExpiration(ctx, h, *opts.NoExpiration))
-	}
-	if opts.AddHours > 0 {
-		catcher.Add(m.extendExpiration(ctx, h, opts.AddHours))
-	}
+	catcher.AddWhen(opts.InstanceType != "", m.setInstanceType(ctx, h, opts.InstanceType))
```
[the remark this answers is not in the document]
Comment (tzembo) [reply cloud/ec2.go]: Agreed. Also, AddWhen seems to evaluate the error in the second parameter even when the condition in the first isn't met? This was breaking the cloud tests.

File: cloud/ec2.go (hunk 6 above)
Comment (tzembo) [reply cloud/ec2.go]: Made a FindHostWithVolume function in the host package that takes a volume ID. Added a check at the host modify rest route that the volume is not already attached to a host.

Comment (tzembo) [reply cloud/ec2.go]: Although I'm now realizing that querying just on the volumeIDs will also look through terminated hosts -- need to investigate what happens to volumes when hosts are terminated and if we should remove volumeIDs from hosts at termination.

Comment (tzembo) [reply cloud/ec2.go]: Decided to remove volumes from host document at termination -- problem should be solved.

File: model/host/db.go [hunk 9]
```diff
@@ -976,3 +981,13 @@ func AggregateLastContainerFinishTimes() ([]FinishTime, error) {
 	return times, nil
 
 }
+
+// FindOne gets one Volume for the given query.
+func FindOneVolume(query db.Q) (*Volume, error) {
```
Comment (tychoish) [inline model/host/db.go]: would it be easy to remove db.Q from the interface?

File: rest/route/host_spawn.go [hunk 10]
```diff
@@ -344,6 +367,142 @@ func (h *hostStartHandler) Run(ctx context.Context) gimlet.Responder {
 	return gimlet.NewJSONResponse(struct{}{})
 }
 
+////////////////////////////////////////////////////////////////////////
+//
+// POST /rest/v2/volumes
+
+const (
+	volumeDefaultProvider = evergreen.ProviderNameEc2OnDemand
+	volumeDefaultRegion   = evergreen.DefaultEC2Region
```
Comment (tychoish) [inline rest/route/host_spawn.go]: why create these aliases?

File: rest/route/host_spawn.go [hunk 11]
```diff
@@ -344,6 +367,142 @@ func (h *hostStartHandler) Run(ctx context.Context) gimlet.Responder {
...
+)
+
+type createVolumeHandler struct {
+	sc data.Connector
+
+	Type string `json:"type"`
+	Size int    `json:"size"`
```
Comment (tychoish) [inline rest/route/host_spawn.go]: I'm a little weird about serializing directly into the handler, and it seems like we'd probably want to have this be the same structure that we use to report the volume.

File: rest/route/host_spawn.go [hunk 12]
```diff
@@ -344,6 +367,142 @@ func (h *hostStartHandler) Run(ctx context.Context) gimlet.Responder {
...
+		Type:      h.Type,
+		Size:      h.Size,
+	}
+
+	// TODO: Allow different providers/regions
+	mgrOpts := cloud.ManagerOpts{
+		Provider: volumeDefaultProvider,
+		Region:   volumeDefaultRegion,
+	}
+	mgr, err := cloud.GetManager(ctx, mgrOpts, evergreen.GetEnvironment().Settings())
```
Comment (tychoish) [inline rest/route/host_spawn.go]: I think there's an env that you can pass into the route construtor to avoid doing the getenv/settings dance.

**Comment by tzembo** [comment]
Here's the state of things on the last day of my rotation:

I created a new VolumeAttachment struct in the host package, a slice of which is stored in the host document in the Volumes field. Each one contains a VolumeID field and a DeviceName field. This kind of deprecates the old VolumeIDs field in the host documents and allows Evergreen to better keep track of the device name associated with each volume.

With this new struct, it should be easier to validate the device name specified by the user when attaching volumes using the Volumes field of the struct and the guidelines here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html (we should loop through the possibilities and assign the first device name that isn't taken!).

**Comment by ablack12** [comment]
Closing this pull request since there's a new one open.

**Pull request closed without merging**